### ISSUE TITLE
Rethinking the embed widget

### ISSUE THREAD

**Issue by jwirfs-brock** [denverfoundation/storybase#741]
## Requirements
- All widgets should have a link to the Floodlight Homepage
- There should be a version of the widget widget that shows a specified featured story
- There should be a version of the widget that shows a list of stories selected by a Topic, Keyword, Project, Organization or Place (originally specified in #777, #779)
- There should be a version of the widget that shows a specified featured story _and_ a list of stories selected by a Topic, Keyword, Project, Organization or Place (originally specified in #778) 
- Clicking the story title link in the widget opens the story viewer in a popup window (originally specified in #775)
- "Cookbook" of example widget markup to achieve different layouts/types of lists
- Embed code will be moved out of the "share" button/widget and into a specific "Embed" button (originally proposed in [this comment](https://github.com/PitonFoundation/atlas/issues/741#issuecomment-18986391))
## Pending design questions
- Do we add a link to the story detail page anywhere in the widget? If not, should we add a link back to the detail page in the viewer?
- What is the default embed code for the "embed" button?  I propose the single story for the story detail page and the project/organization list for the project/organization detail page
- How do we plan to depricate the existing widget? That is, do we switch existing widgets to the new one automatically, or maintain the old one.  How do we differentiate between widget versions?
## Action items
- Tests for updated widget (#819)
- Define "syntax" for widget configuration (#820)
- Templates/CSS for widget layouts (#821)
- Django view for updated widget (#822)
- Update embed widget JavaScript (#823)
- Embed button for story detail, project detail and organization detail pages (#824)
- "Cookbook" of different embed widget configurations (#825)
## Original issue description

In the conversations we've had with Piton and Denver Foundation about refining the focus and incorporating Floodlight into the grant-making process, the theme of **embedding Floodlight stories on external websites** has come up several times. I've pointed out that we do have a simple embed widget (to try to understand what featured people would like to see that don't exist now). Here are some of the main issues:
- Users would like to have an embed widget that contains more than one story (i.e., a stream of all of their stories, or a stream based on a location or tag).
- Users would like to have some basic branding/styling of the widget, so that it "fits" with their website (could be a logo, could be styling?).
- Users (and **especially non-profits**) want to keep people on their website rather than sending them to an external site. They'd like to see a way to read the **full story** via the embed widget.
- We are still working out ways for users to easily contribute to an existing story, but once we have that, it should be incorporated into the embed widget (i.e., readers should be able to contribute without having to go back to the Floodlight site).

I've sketched up a couple of options that address some of these issues (GitHub is doing some weird scaling, so they aren't supposed to be huge).

**Widget layouts that have multiple stories**
A. Vertical layout, main story + list
![fl_embed_expanded](https://f.cloud.github.com/assets/1020518/466067/86c00d62-b64f-11e2-98c0-a937e6edc2a7.png)

B. Horizontal layout, main story + list
![fl_embed_horizontal](https://f.cloud.github.com/assets/1020518/466068/8d57cab6-b64f-11e2-96b6-513fdee381cc.png)

C. Slider layout
![fl_embed_slider](https://f.cloud.github.com/assets/1020518/466070/9551ca0a-b64f-11e2-9f8c-efb81cce616c.png)

**Widget layouts that incorporate an "add your voice" feature**
A. Vertical layout with "add your voice" buttons
![fl_embed_addyourvoice1](https://f.cloud.github.com/assets/1020518/466072/9c4ac35c-b64f-11e2-9edc-b158371a6208.png)

B. Expanded "add your voice" button 1
![fl_embed_addyourvoice2](https://f.cloud.github.com/assets/1020518/466074/a24c910e-b64f-11e2-847c-76ce53c2ac71.png)

C. Expanded "add your voice" button 2
![fl_embed_addyourvoice3](https://f.cloud.github.com/assets/1020518/466075/a7d9fbf2-b64f-11e2-8ebb-8f5039171dc7.png)

I haven't yet come up with any sketches that let you read a full story in the embed widget. Ideas on that? In our original widget design, we had talked about having it open up the story viewer in a modal window. @ghing There were technical reasons we couldn't do that, right? Are there other thoughts as to how we might do it? With our new story viewer design, we may have some flexibility with this.

Some other questions we need to think about as we redesign the embed widget:
- How will users choose what goes in the widget (i.e., selecting a project, organization, author, tag, location, whatever that will provide the feed)?
- Will the embed widget be a live feed (like a Twitter feed), or will it be static?
- Will users get to set the featured story?
- Will users get to choose from multiple layouts to find the one that best fits in their website?
- What type of branding/styling will be available? These sketches just have a logo (plus the Floodlight logo) -- do we want anything else beyond that?
- What types of meta buttons need to go on there (i.e., links back to Floodlight)

Some of these questions can be answered by going back to our power users (TDF and Piton). We can also talk to a few other users (i.e., the DPS Americorps project) to see how they might use a beefed-up embed feature.

**Comment by ghing** [comment]
@jwirfs-brock I like your sketches and feel like they're generally in the right direction, but as we/re thinking about embedded widgets, and this issue is being driven by the partners, _we really need to think about this in the context of where the widgets will go on the partners' pages_.

Can you talk to Piton/TDF and get some examples of page contexts where they'd like to see widgets?

I'll do some experimentation with opening a modal. Off the top of my head here are the issues:
- We can't create a modal div of the story from within the IFRAME because of cross-domain issues.  But, because we're creating the IFRAME dynamically via a script, we can probably add functionality to the script that can manipulate the DOM of the page that includes the script. But ...
- Adding logic to the script that handles modal window creation and sizing would increase the payload of the widget script. I need to think a little bit more about what the implications of this are for the partners' page load times.

Beside that, I'm concerned that popping up a modal when clicking on a link is sort of annoying and unexpected for users. This is the behavior that occurs with link spam.  Ultimately, we can try it a couple of different ways and see what users and the partners like.

An idea that I had was to have one widget for the feed, that would link to the stories' detail pages and another widget, similar to the YouTube embed, that would show the story viewer in an embedded space with the option of popping it out into a larger modal or a fullscreen view.

Finally, to get a sense of what user expectations might be, I tried out the Facebook and Twitter widgets on http://www.civiclab.us/ a site that uses both:
- With the Facebook widget, links open in a new tab.
- With the Twitter widget, links inside tweets are opened in a new tab. The link for the user's handle opens up the a smaller version of the user's timeline in a pop-up. The pop-up is a new browser window (without all the chrome) and not a modal overlay.

**Comment by jwirfs-brock** [comment]
@ghing Thanks for the comments. Definitely agree on talking to Piton and TDF. I sent them a survey this morning that looks like this: https://docs.google.com/forms/d/1-OBHaj6FifhUAQuyHlPJMWveDAGi3JGm7mXG2MfADmA/edit#

(I already have a response from Matt, still waiting on Rebecca.)

As for popping up a model window, I agree that it's problematic. I like the idea of having the story viewer show up in the embed widget itself. But the problem with that is, how do you size the widget? If it's small enough to appear nicely on someone's webpage, the viewer will probably be too small to read. It's something to chew on a bit more...

**Comment by jwirfs-brock** [comment]
@ghing Oops, we have Rebecca's response too. Yay!

You can view them here: https://docs.google.com/spreadsheet/ccc?key=0As2exFJJWyJqdHVZaXdMSzl3VlpNQlYwMWROeHdkNXc&usp=sharing

I forgot to add a "Who are you" question. Piton is the first response, TDF is the second.

**Comment by ghing** [comment]
I did some experimenting with a version of the embed widget that opened the story in a modal viewer. I was able to:
- Add jQuery and SimpleModal to the parent page dynamically
- Create an iframe with the story viewer and add it to the parent page

I wasn't able to wire up click handlers to the widget iframe (what we have now) because of cross-domain issues. We could get around this by writing the widget HTML straight to the DOM instead of loading it in an iframe, though that makes the widget JavaScript bigger and more complex. I'm going to hold off on experimenting with the modal further (partially because I think this behavior might be kind of annoying) and try opening the viewer in a popup.

**Comment by ghing** [comment]
Another question, with the new widget, do any links go to story detail pages or do they all just open the viewer? I'm not sure if we want to totally abandon directing users to story detail pages, mainly because they provide links for users to explore floodlight and see other stories. We could add navigation in the viewer that would close the viewer popup and direct the user to other Floodlight pages, so this is more of a consideration than an actual problem.

**Comment by ghing** [comment]
http://www.wbez.org/news/education/zero-trust-after-cps-admits-it-overstated-savings-closing-schools-107044 and most stories on http://wbez.org are an example of why I'm reluctant to open up the modal popup viewer.

Granted, we'll be launching the modal on a click-event, which makes it a little more expected, but I'm a little concerned about making behavior that feels like annoying ads.

**Comment by jwirfs-brock** [comment]
@ghing Point noted about WBEZ. You are right that model typically reads as "ad" -- and that's not the message that we want to send.

In response to the question, "do any links go to story detail pages or do they all just open the viewer," I think we'll need to look at the feedback we get from Matt, Rebecca and other users on how they would like to use the embed widget. If they just want to use it to feature their own content, then it makes sense to only show the viewer (and not the story detail page). If they want to feature Floodlight content as a way to show that they are part of a larger community of content contributors, then it makes more sense to show the story detail page. My initial instinct was to just show the story viewer, but I think this is something that will require a bit of analysis.

**Comment by ghing** [comment]
@jwirfs-brock Agreed about waiting to synthesize feedback from partners about link targets. My feeling is that if links open the viewer, we should at least have one link in the widget that links back to the Floodlight homepage and also a link in the viewer that allows the user to get back to Floodlight if they want.

**Comment by jwirfs-brock** [comment]
@ghing Based on the early feedback from Matt and Rebecca, the one thing that is clear and consistent is that they would like readers to be able to view the full story content on their websites without having to go back to Floodlight.

To accomplish this, viewing the story in the bounds of the embed widget is a possibility, and would work as long as the embed widget is large enough. For example, we could accomplish this with (B) above by, when a user clicks on a story to read, opening the viewer in such a way that fills the entire widget (while still offering options to close the viewer and go back to the initial view and to view the story on Floodlight).

For some of our stories, they would look nicely in this size frame. For others, they have huge pictures or huge blocks of text. So for this to work, we would have to do a better job of constraining the content in such a way that it would be viewable in our widget. In some ways, this could be an advantage, although I'm sure it will pose some exciting challenges.

Another option -- and this could be really crazy and impossible -- would be to have the widget **look** like it was expanding to take up more of the page when you launch the viewer in the widget. In a sense, this would be like adding a popup/modal that doesn't look like a popup/modal, it just looks like the widget is sliding over to expand.

Here's a picture of the idea...

The Denver Foundation is thinking of embedding Floodlight content on their topic area pages, like this one: http://denverfoundation.org/community/page/supporting-education-1

Here's what it might look like to have a widget on that page:
![df_embed1](https://f.cloud.github.com/assets/1020518/483135/a6b6a9d0-b8b3-11e2-9ef0-dbca6daabff0.png)

And here's what it might look like to "expand" the widget to open up the viewer:
![df_embed2](https://f.cloud.github.com/assets/1020518/483136/ada70302-b8b3-11e2-81d8-e56ff9947835.png)

I'm probably not doing a very good job of describing this, but instead of clicking on a story and seeing a popup (that could potentially be mistaken for an add), it would like like the widget is just sliding over/expanding as the content inside it changes to the story viewer.

**Comment by ghing** [comment]
@jwirfs-brock, so one big takeaway from this is that, at least for the Denver Foundation, the space for the widget wouldn't easily accomodate the story viewer.

I think I get what you're going for in terms of the slide effect of the viewer from the widget.  I can experiment with that when I have time.  I'll have to do more thinking about this, but I'm pretty sure that in order to do this, we'd have to write the widget HTML directly to the DOM of the parent page (instead of being shown in an iframe).  This would make all the elements much more subject to the parent site's CSS.  We may ultimately have to consider a custom widget for key partners and a less fragile widget for general consumption.

**Comment by bpawlak** [comment]
My thoughts, kind of in summation of what I've heard above:

1)  I like the idea of putting some parameters around the embedded story.  Users could pick a single story, click "embed this story" and then get a wizard that leads them through the rest of the options (see below).  Pick 1 main embedded story, option to include 3 "sidekick stories" that are selected based on tags, location, organization, etc. change some basic CSS properties, etc.

2) I don't see the point of co-branding, given the amount of extra work it would take (logo sizes, managing uploads, file formats, different sizes based on widget, etc.)... the widget will be embedded on another site which will already have a logo and brand and content, etc.  Is the plan to enable Organization A to embed the widget on Website B so that Organization A's logo is shown on Website B?  I think this is a Floodlight embed widget and, as such, should be viewed as a value-add offering from Floodlight (and therefore another way to promote Floodlight). I don't think there should be options to a) get rid of the floodlight branding or b) get rid of the link to the Floodlight website.  If the Floodlight website doesn't get traffic and contributions, there won't be anything to embed!  Like Geoff said, we want to make sure people can get to the Floodlight website (to read stories, create new, contribute to a story, etc.)

3) I don't agree with the ask to contribute to a story from the widget.  We want people to _read_ the story before adding to it, so the "contribute your voice" should come as part of the story viewer.  Makes the widget cleaner as well.

4) I think it's worth trying to figure out how to show a viewer pop-up, but couldn't we start with just opening another tab?

![embed_wizard](https://f.cloud.github.com/assets/2250519/490149/cc4eb972-b9b1-11e2-8282-5b2ffa6b40d8.png)

**Comment by bpawlak** [comment]
I can build out the panels to the wizard as we discuss functionality, etc.

**Comment by jwirfs-brock** [comment]
@bpawlak 
1) I like the idea of centering widget creation around a story. But we do definitely have interest from Piton and TDF around embedding streams. I'll have to hear a little more about what you mean by "sidekick" stories to see if this will meet their needs.

2) Totally agree with you on this one. If it's embedded on their site, they don't need to have the logo inside the widget. Also agree that we need to keep the Floodlight logo in there and link back to the site (and have a prominent "Read this story on Floodlight" link).

3) This is the one point where I disagree. I think that we really want people to be pulled in and feel like it is a participatory widget. I think the "Add your voice" prompts can be framed in a way that it's a question people can respond to whether or not they have read the story (like, "What do you love about your neighborhood?"). This (a) makes it really easy to contribute in a meaningful way, and (b) further distinguishes the "Add your voice" feature from commenting. But this is an area that still requires a lot of prototyping and testing, and I could be totally off the mark.

4) Totally agree. That is what the widget does now, I believe.

**Comment by bpawlak** [comment]
1) Yes, I was thinking the "organizational embed" would be similar in that it would be a wizard with steps to configure, but it wouldn't necessarily have to start with a story.  It'd mostly just start with Step 2 (although it wouldn't be called "sidekick stories" - that concept is just a way for a user to specify which other stories they would like to embed along with the main story they're embedding.).  The "organizational embed" would only require the user to set the parameter/tag to grab for the stream of embedded stories.

3) Will people need to have a Floodlight account to contribute to the story?  I'd hate to try to force registration, etc. through the widget.

**Comment by ghing** [comment]
### Re: wizard

@bpawlak, is there any reason to not just have all the wizard steps compressed into a single view? In particular, I think it's helpful to be able to see the widget change as you change options (particularly color/size ones).  Also, most other widget generation steps I've encountered do it this way, which isn't necessarily a reason to follow their lead, but I'd be interested in knowing if you find these implementations lacking

### Re: story curation, "sidekick" stories

@jwirfs-brock For the mockups you made, do you imagine that the user will select the larger story and that the other stories will be auto-populated from the most-recent stories from an organization? Or is the user selecting all the stories that appear in the widget?

### Re: contributions to story inside the widget

@jwirfs-brock I think that for thinking about story contributions, we need to work backward from how the contributions will be presented in the story viewer of the seed story.  I think we all agree that we want reading a story to be participatory and we want to offer a spectrum of ways of responding to/contributing to the story, but I feel like we haven't figured out what this looks like. It might be early to try to think about what the interaction looks like in the widget when we're not sure what the interaction is.

@bpawlak I totally agree with this:

> I'd hate to try to force registration, etc. through the widget.

**Comment by bpawlak** [comment]
I think we can have a real-time view of the widget and how it will present itself.  That was essentially my thinking for Step 3... allowing users to dynamically change options on the fly and see how things would look.  I guess we _really_ don't need step 4 - it was just intended to give a "clean" view of the widget without any of the controls used to modify parameters/settings/etc.

We _could_ try to integrate Steps 1 & 2 into the real-time preview as well, but as with the tagging, location and featured image part of the story builder, putting in too much functionality into one screen might be confusing to our users.

**Comment by bpawlak** [comment]
Here's a summary of what Jordan and I discussed today and taking into account Geoff's suggestion in [https://github.com/PitonFoundation/atlas/issues/741#issuecomment-17813632].

![embed_dialog](https://f.cloud.github.com/assets/2250519/608660/26dc32ea-cd65-11e2-93dd-c9561b076f53.png)

**Comment by ghing** [comment]
@bpawlak and @jwirfs-brock, the mockups of the wizard look good to me.  A few questions/comments:
- How will the user launch the wizard from the share popup? Will the wizard open in a separate tab or in a modal popup?
- It seems like before we implement the wizard, we should implement the new widget styles themselves. Has there been any clarification on the design of the widgets themselves? I get a pretty good idea from the options in the layout, but I could use some clarity.
- How many related stories should be shown? Can the user select multiple tags?

**Comment by bpawlak** [comment]
Oops.  I forgot to follow up the wireframe post with some more details... sorry about that.

To answer the 1st & 3rd questions:

1) The "Embed this story on another site" line in the share popup should be changed to a button.  Clicking on the button would close the share popup and show the widget creation tool as a modal.  One thought I had is that perhaps the "Embed" should be it's own button so that there would be two: 
![shareembed](https://f.cloud.github.com/assets/2250519/612826/2bf039c6-cdf5-11e2-8c5e-63396e4fc919.png).  

That way, on an organization (or project) page, we'd have a button to "Embed this $ENTITY's Stream" which would automatically set the tag in the widget creation tool.  I'd think we might want to differentiate the Embed and Share buttons using icons or colors, but let's discuss:

![icons](https://f.cloud.github.com/assets/2250519/613020/dd8569c0-cdf7-11e2-8aa7-3085a7324f0a.png)

3) I was initially just assuming a max of 3 related stories based on conversations we've had, but of course that's easy to change to a user control.  The downside is that if we enable an arbitrary # of stories, we'll have to  worry (more) about scrolling, sizing, etc.  I was also thinking that it would be limited to a single tag as there's then no confusion about AND or OR between multiple tags and it also makes the auto-complete that much simpler.  Again, just my initial thoughts for a V1 release.

One other thing that came up is that I think it's clear from the users we talked to that they'd like the story viewer to appear as if it's part of the site in which it's embedded:

![read_story](https://f.cloud.github.com/assets/2250519/612933/d7dadc0e-cdf6-11e2-914a-b99cdeaf977e.png)

I know you said that might require some additional research, so I think you can go ahead and start that effort.

**Comment by ghing** [comment]
@bpawlak,

> I was also thinking that it would be limited to a single tag as there's then no confusion about AND or OR between 
> multiple tags and it also makes the auto-complete that much simpler.

I think a single tag is better too. I think the purpose of the widget gets murky otherwise.

> One other thing that came up is that I think it's clear from the users we talked to that they'd like the story viewer 
> to appear as if it's part of the site in which it's embedded

Any examples of this out in the wild would be appreciated.

I'd prefer to continue to embed the widget in an iframe (to sandbox widget styles and to leverage server-side templates to make rendering or updating the widget easy), but, as far as I can tell, that makes it tricky to break the viewer modal out and dim the background elements. We can do what Twitter does and open a new window (see https://dev.twitter.com/docs/embedded-timelines). 

If the popup isn't acceptable, one (untested) idea I just had would be to add click handlers in the parent DOM for the links in the iframe that launch a modal in the parent DOM.  I think that should work.  The downside is that it makes the widget script payload significantly bigger because you have to include code to manage the modal.

**Comment by bpawlak** [comment]
> > One other thing that came up is that I think it's clear from the users we talked to that they'd like the story viewer 
> > to appear as if it's part of the site in which it's embedded
> 
> Any examples of this out in the wild would be appreciated.

Once we build it, I can show it to you.  :)  

I think Jordan has found out that opening in a new window (as it works now) isn't as desirable to our potential users, although I personally think it's a good first step.

**Comment by ghing** [comment]
> I think Jordan has found out that opening in a new window (as it works now) isn't as desirable to our potential users, although I personally think it's a good first step.

By "as it works now", do you mean in a new tab (or whatever the user's preference is), or a popup like in the [Twitter example](https://dev.twitter.com/docs/embedded-timelines) I mentioned before? I had a Twitter-like popup working a as a demo for a while on dev before I switched gears to work on #583.

**Comment by bpawlak** [comment]
I only saw it in the new tab, didn't see the popup.  Need to get Jordan's clarification on what the requirement is.

**Comment by ghing** [comment]
To me, the ideal implementation would be to open in a popup (like Twitter) by default.  Then we could include an additional script that would set up click handlers to open the viewer in a modal if desired (thus keeping the default widget script lightweight).

Alternately, what about switching to fullscreen mode, a la the [HTML5 version](http://www.youtube.com/html5) of the [embedded youtube player](http://apiblog.youtube.com/2010/07/new-way-to-embed-youtube-videos.html)?

Note to self: http://blog.johnmckerrell.com/2006/10/22/resizing-iframes-across-domains/ has some good ideas for resizing iframes across domains.

**Comment by jwirfs-brock** [comment]
@ghing I'm not quite sure what you mean when you refer to the Twitter example -- when I look at that page, I just see the embedded timeline, not a pop-up window. And then when you click on the "Tweets" link, it opens up the Twitter profile page in a new tab.

@ghing and @bpawlak In general, I'm having a little trouble following the options and understanding what's possible and what isn't. 

So, just for my own sanity, the options are:
- "Click to read story" link opens in new tab/window (easy -- this is how we implement it now)
- "Click to read story" link opens viewer as a pop-up (easy, but readers might find it annoying)
- "Click to read story" link opens a full-screen player like YouTube (difficulty? Bill and I discussed this a little on Tuesday and weren't sure how we would get the scaling to work)
- "Click to read story" link opens a modal window, like what happens on the Floodlight site (this is technically hard)

Is that right, or did I miss something? It would be helpful for me to flesh out the pros/cons of each option.

It isn't necessarily essential that the story viewer pop-up within the current page. I think we can make it work by opening a new tab, as we do now. When TDF and Piton indicated that they'd like users to be able to read stories at their site, the implication was that **the widget needs to feel integrated into the site and reading stories that are contained in it doesn't feel awkward.** 

@bpawlak I really like the suggestion of breaking out the embed button from the share button. That definitely improves the workflow.

**Comment by jwirfs-brock** [comment]
@ghing @bpawlak 
Here's an example that shows why it is probably ok to have the embed widget open in a new window or tab. The NY Times often has multimedia extras that go along with their stories (like in this one: http://www.nytimes.com/2013/06/02/magazine/how-chicagos-housing-crisis-ignited-a-new-form-of-activism.html?ref=magazine). When you click on the extra - whether it's a slide show, like in this example, or an interactive - it opens a separate view, but isn't super jarring. I don't _love_ it - there is still a little bit of a disconnect - but it doesn't completely annoy me as a reader because I'm used to it. The analogy here is that someone's website (example, TDF) would be the main story, and the Floodlight story would be the multimedia extra.

**Comment by ghing** [comment]
> @ghing I'm not quite sure what you mean when you refer to the Twitter example -- when I look at that page, I just see the embedded timeline, not a pop-up window. And then when you click on the "Tweets" link, it opens up the Twitter profile page in a new tab.

@jwirfs-brock, in the embedded timeline widgets, click on any of the usernames in the timeline. For me, it opens the user's timeline in a popup.

> "Click to read story" link opens a full-screen player like YouTube (difficulty? Bill and I discussed this a little on Tuesday and weren't sure how we would get the scaling to work)

Not too difficult since there's an API that provides fullscreen functionality in modern browsers. The tricky part is supporting Internet Explorer. In this case, we could do what YouTube does and just make the video 100% of the browser window (as opposed to 100% of the screen).

What challenges did you see with scaling? To me this seems easier than calculating the margins/offset/padding to properly center a modal popup.

> "Click to read story" link opens a modal window, like what happens on the Floodlight site (this is technically hard)

I'm sure we can figure out a way to do this well, but I think it will take a lot of refinement to get it working well across browser and in different embedding contexts. I want to make sure we don't get bogged down in an interaction that doesn't add much additional value.

The other downside to this is the script that must be downloaded into the users browser will be much bigger because it has to contain logic for sizing, managing and positioning the modal popup. We'll also can't rely on easy tools to do this like simpleModal unless we want to force downloading even more scripts and figure out how to reuse jQuery already on the page or avoid jQuery version conflicts.

Once I finish up the asset dataset stuff, I want to experiment with a strategy that I think will work for doing this in a way that has a simple default behavior and then can add more complex behavior by including an additional script.

> When TDF and Piton indicated that they'd like users to be able to read stories at their site, the implication was that the widget needs to feel integrated into the site and reading stories that are contained in it doesn't feel awkward.

I need a bit more granularity on this. I want to do some early prototyping to make sure Piton/TDF are happy with the direction before spending a lot of time refining things. For instance, maybe they want to be able to tweak the colors/graphical elements of the widget. Do we build a styling editor into the wizard? Or maybe careful design of the markup and a cookbook of CSS styles will give them exactly what they want without building a feature that only gets used a few times.

**Comment by jwirfs-brock** [comment]
@ghing I agree, I do think we need some clarification from TDF and Piton in terms of WHY they said they would prefer to have people be able to read stories from their websites.

The exact way I worded the question to them before was: 
"How should readers be able to view the Floodlight story/stories on your website?"

Everyone chose: "...read the full story on my website without having to leave" over "...read a summary of the story on my website, then to to Floodlight to read the full version"

**Comment by ghing** [comment]
@jwirfs-brock Re: the NYT example, when I click on an item like one of the slide shows in the story you pointed me to, it actually replaces the page in my current browser tab, which is the normal link behavior. Personally, I like this, and a lot of people argue that you should just make links work the way they're intended and leave it up to the user to decide whether to open it in a new window or not.  Unfortunately, years of people trying to force behavior on users has made it really hard for users to know what to expect when they click a link. We have to untangle this mess the best that we can.

Maybe we need a mission statement for this feature to guide us?

**Comment by bpawlak** [comment]
I agree with you, Geoff, and I think Jordan does, too.  "Standard" link behavior just replaces whatever page I'm on with whatever the target page is.  If a user wants to open a link in a new window, they can do so (unless it's asp.net or some javascript trying to prevent it (grr!)).  I think Piton & TDF are worried that if people leave their sites to go read a story, they won't ever come back to the original site.  That's a pretty dated concern for websites, in my opinion (users are going to do what they want to do and we shouldn't try to force them to do something other than that).

I think we should just pick something and build it.  In Jordan's options, above, we already have the first option working, but that apparently isn't ideal from Piton & TDF's perspective and admittedly it feels a bit unpolished.  

So let's try the 3rd option: "A 'Read Story' link (shown with a "fullscreen" icon (e.g.,http://www.iconfinder.com/icondetails/114961/128/bigger_fullscreen_make_icon) opens a full-screen player like YouTube.  The text and images scale appropriately and pretty much everything is the same with the viewer only the "close" icon closes the fullscreen view instead of closing the viewer modal as on the floodlight site.

**Comment by jwirfs-brock** [comment]
Main takeaways:
1) **Our next iteration of the widget doesn’t need to include a wizard.** First, we should focus on building the widget so we can demonstrate the different options. We can offer support to some power users (TDF, Piton, DPS Americorps, Denver Public Health) to help them customize the features.
2) **Clicking on a story link within the widget should open up a pop-up window that displays the story in the story viewer.** (Similar to Twitter’s embedded timeline feature: https://dev.twitter.com/docs/embedded-timelines)
3) **Three widget formats to try: featured story, list of stories, featured story + list.** We don’t yet know whether widgets will be sidebar content or in the main body of the page. We can play with this.
4) **Open projects (a project that any user can add a story to) is higher priority** than customizing the widget story feed options.

Other notes: 
Example use cases (from Matt):
1) _Embedding stories about Piton focus areas_ - They would have a neighborhood-based initiative that has a page within Piton’s website. That page gives (Piton-authored) context about the initiative. It could include a Floodlight widget of user-contributed stories related to the initiative (created via tags, stories that are part of a project, etc.)
2) _Prize philanthropy_ - They would offer a prize for the best story that illustrates work in a certain area. People could enter the contest by authoring a Floodlight story (that is part of a particular project). That project/contest feed would then be embedded as a list (with no featured story) on Piton’s website.
- May potentially have more than one widget on a single page
- Would want to be able to search Floodlight to create a custom feed of stories (i.e., checking off which five stories you want, or picking a tag, or a geography, or a project or an organization...)
- Want to be able to share any stories directly from the window you are viewing them in (i.e., from the pop-up you get from clicking on a story link in the embed widget)

Questions:
- Do people really want to embed stories, or do they want a really nice-looking badge/button they can display on their site that links back to Floodlight?
- How does the way an organization who is using Floodlight has part of their communications strategy use a widget contrast with how I-News (or other news outlets) would use it?

Next steps:
- Jordan will initiative conversations with other power users (DPS Americorps, Denver Public Health, CLiC, I-News) to show them the sample widget and get some feedback.
- Geoff will mock up some examples of the widget we can show parters (incorporating pop-up window, list view)

**Comment by ghing** [comment]
Here are wireframes of the new widget designs from @bpawlak:

![embed_options](https://f.cloud.github.com/assets/110420/865351/44388cda-f637-11e2-9349-bfab4dedfb96.png)

I feel like we shouldn't implement the carousel view initially due to the additional styling and JavaScript required to implement this.

Also note that there isn't a wireframe for the list-only version, but it should be possible to extrapolate this from the featured story + list example.

**Comment by ghing** [comment]
@jwirfs-brock, I noticed that in [the wireframe](https://github.com/pitonfoundation/atlas/issues/741#issuecomment-21648192) for the single story view, there is a link at the bottom to the project or organization for the story.  For a list of stories, the behavior of this link is clear. It should just link to the item's page.  However, for a single story, the desired behavior is less clear.  The link could be:
- The first organization of a story, alphabetically
- The first project of a story, alphabetically
- A project or organization configured by the user via markup. (Though this makes the markup a bit more complicated and doesn't give us a default behavior to easily throw in the "embed" button)
- The first organization and project of a story, alphabetically
- All organizations of a story
- All projects of a story
- All organizations and projects of a story
- The story detail page (where they'll be able to see all the story's projects and organizations)
- omitted altogether

What should be the behavior of this link

cc @patternleaf

**Comment by jwirfs-brock** [comment]
@ghing I think it makes sense to list all organizations and projects the story is in (if there are any), similar to how a story's affiliations are listed on the story detail page.

**Comment by ghing** [comment]
@jwirfs-brock, Ok.  What would the microcopy for that look like?

Were you thinking something like this?

_View all stories for Project1, Project2, Organization1_

**Comment by jwirfs-brock** [comment]
@ghing Yeah, I like that. We might want to test it with some of the stories affiliated with longer project and organization names (like the [DPS ones](http://floodlightproject.org/en/stories/americorps-attendance/)) to make sure the formatting looks ok.

**Comment by bpawlak** [comment]
I thought that the embed widget would only have stories from a single project or a single organization?  When would we have "View all stories for Project1, Project2, Organization1, etc.?"

**Comment by jwirfs-brock** [comment]
@bpawlak It's not that common, but some stories are in multiple projects, or associated with multiple organizations.

**Comment by bpawlak** [comment]
Ok.  Thanks for the clarification.

**Comment by ghing** [comment]
From @jwirfs-brock, suggested example stories for testing:

> For testing multiple story widgets...
> - Take a stream of the four latest stories in the "Denver Public Schools AmeriCorps" project:  http://floodlightproject.org/en/explore/?projects=c0886c31cca444c78c0a90f070fb4694
> - Use the four stories in the "Nonprofit Internship Program of the Denver Foundation" project: http://floodlightproject.org/en/explore/?projects=e01e7cc0ad714a43885aa617f3a09222
> 
> For testing single story widgets...
> - "Why Libraries Are Relevant" http://floodlightproject.org/en/stories/why-libraries-are-relevant-2/
> - "Timeline: Civil rights progress, defeats since 'I Have a Dream'" http://floodlightproject.org/stories/timeline-civil-rights-progress-defeats-since-i-hav/
> - "Meet (a few of) the Artists from Redline" http://floodlightproject.org/stories/meet-a-few-of-the-artists-from-redline/
> - "SalusWorld Heals Trauma Scars from Within" http://floodlightproject.org/en/stories/salusworld-heals-global-trauma-scars-from-within/

**Comment by ghing** [comment]
@jwirfs-brock for the multiple story widgets:
- what's a good default number of stories to show?
- should the user be able to configure how many stories show up in the widget?

For the story + list widget, how many stories should we show (the wireframes show 3)?

What are good test examples for the story + list widget?

**Comment by jwirfs-brock** [comment]
@ghing Here are some answers to your questions

> what's a good default number of stories to show?

I like the way Bill set it up in the wireframes with a main story plus three smaller stories

> should the user be able to configure how many stories show up in the widget?

To avoid adding additional complexity, let's just standardize it for now.

> For the story + list widget, how many stories should we show (the wireframes show 3)?

Yeah, I was just going on the wireframes for this. Let's stick with it.

> What are good test examples for the story + list widget?

The first two example projects I gave are good ones to use for this. How about use the most recent story in the project as the main story, with the others as the smaller/"sidekick" stories.

**Comment by ghing** [comment]
@jwirfs-brock the way we had been talking about it, we talked about having 3 widgets types:
1. Primary Story only (pictured in wireframes)
2. Primary Story + List (pictured in wireframes)
3. List (not pictured in wireframes. I imagined this would look like the right hand side of the story + list wireframe)

Do you still want to have a list-only widget, or if the widget should always show a Primary story.  If the widget should always show a primary story, should the user always specify this in the widget markup, or should this default to the most recent story in the organization, project, etc. if the user doesn't provide it?

**Comment by jwirfs-brock** [comment]
@ghing Yes, let's still have a list only widget.

For the primary story + list widget, will we still  be able to set up this by choosing a project/organization/user as the source of the stories? If so, then I would say let the user specify the featured story in the widget markup, then if no featured story is specified, default to the most recent story.

**Comment by patternleaf** [comment]
@bpawlak @jwirfs-brock @ghing Do we have some expected/known dimensions or ratios to work with? Currently I'm shooting for reasonably responsive but some ratios are definitely more natural than others given the above wireframes.

**Comment by bpawlak** [comment]
There was a an email thread where I took a look at our most likely candidates for using the embed widget (and where on their sites it'd be used) and trying to determine a general "best fit" for width.  The end result was that the candidate sites all have (predictively) different horizontal layouts in terms of room available for a widget.

Approx sizes:
- i-News (either 600 or 935px)
- CLiC (500, 715 depending on where embed)
- CP (700px)
- America Scores (730)
- DenverDebate (580)
- Earthlinks (670)
- MileHighConnects (700)
- UnitedWay (690 or 510)

At one point we talked about letting the user explicitly specify a width for the width, but the responsive route is far more flexible from the end user's perspective.  Can you use the above as rough guidelines for breakpoints?

**Comment by patternleaf** [comment]
I put together a little testing page for the widget:

http://dev2.floodlightproject.org/static/widget-test.html

You can modify dimensions, story and list sources. Only tested in Chrome. 

As currently implemented, the widget really needs a certain ratio of vertical space to horizontal space to begin to look decent. Even then, lots of stuff tends to get truncated: bylines, images, and so on. There are only two breakpoints right now, and they're not yet tuned or optimized.

Do we have a sense for _vertical_ space available?

**Comment by bpawlak** [comment]
I think we can reasonably assume we have as much vertical space as necessary to show the content.  I know it's not truly unlimited, but I think it could be greater than the 300px you have as default ... if I put in 500px on the test instance you created, then at least I can see all of the story titles.

**Comment by jwirfs-brock** [comment]
Would it be possible to standardize the dimensions of the images, especially for the stories on the right? When I set the height to something large (600 px or greater) the top image (for "Eating Your Own Dog Food") expands to its full dimensions, which means its vertical height appears larger than its width. This looks funny, given the fact that the other images are roughly square.

**Comment by jwirfs-brock** [comment]
Also, a question for the whole team -- any thoughts on whether the story links should go to the detail page or launch the story in the viewer (as it is now)? I know we had reasons for launching stories in the viewer, but looking at it now I think linking to the story detail page might make more sense. Thoughts?

**Comment by bpawlak** [comment]
So _none_ of the links would open the story in the viewer - they would simply take the user to the Story Detail page on Floodlight?

**Comment by jwirfs-brock** [comment]
@bpawlak Yeah, just putting that out there to see what people think. If most of the team likes launching directly into the story viewer, I am ok with it.

I mainly suggested it because when you open a story directly into the viewer (and not the viewer launched from the story detail page), it can feel a little bit disorienting or like being dropped on an island. For example, there's no link back to the story detail page or to Floodlight. You can still get there by truncating the URL, of course, but that's the only way.

**Comment by bpawlak** [comment]
I see the concern. Can't we edit the embedded story viewer (or perhaps simply add it to all viewer instances) so that there's a link to the story detail page (on Floodlight)?

**Comment by jwirfs-brock** [comment]
@bpawlak I think your suggestion would work.

**Comment by patternleaf** [comment]
> When I set the height to something large (600 px or greater) the top image (for "Eating Your Own Dog Food") expands to its full dimensions, which means its vertical height appears larger than its width. This looks funny, given the fact that the other images are roughly square.

@jwirfs-brock We should be able to scale/crop images to maintain consistency, though again it all gets wonky if the vertical dimension of the whole widget is unknown.

If we could standardize on a height for the widget, or a couple of heights, that would be awesome.

(I want [flexboxes](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Flexible_boxes)!)

> Can't we edit the embedded story viewer (or perhaps simply add it to all viewer instances) so that there's a link to the story detail page (on Floodlight)?

This makes sense to me—there's no branding whatsoever on the viewer. @bpawlak Would you like to wireframe what you have in mind?

**Comment by ghing** [comment]
@jwirfs-brock, @bpawlak, @patternleaf I opened #851 as a separate issue for linking back to the homepage and to the story detail page in the viewer.

**Comment by ghing** [comment]
Did a code review today of the widget template and widget-generating JavaScript.  Tasks:

@patternleaf:
- [x] Support list-only widget in storybase_story/story_widget.html or implement storybase_story/story_list_widget.html
- [x] Implement storybase_story/widget_404.html
- [x] Make all story links go to viewer
- [x] Click handler instead of inline call to window.open()
- [x] Click handler will try to call window.open() in a way that opens in a new window rather than new tab
- [x] Wrap strings in {% trans %}
- [ ] Use {% featured_asset_thumbnail_url %} template tag if you need to try to force a particular thumbnail size
- [x] Use {% fullurl project.get_absolute_url %} to get links for taxonomy terms in footer
- [x] Include Modernizr in widget view to support HTML5 tags in old browsers

@ghing:
- [x] see if protocol relative URLs ("//") works from https page and use them if they do. _They don't work. In fact, neither the JavaScript or the iframe will load if the page including the widget is served via HTTPS.  We should get a certificate._.
- [x] get script domain using http://stackoverflow.com/a/2976714/386210
- [x] check if we need to trim url in getUrl(). _We do. Also need to polyfill trim() for IE_
- [x] widgets.js: line 106, 107, make sure an element is found or try/catch for everything inside getUrl. _try/catch seems cheaper, more robust_
- [x] documentation comment for getUrl

**Comment by patternleaf** [comment]
Going to add to my task list:
- [ ] Widget styling in IE8

Currently pretty gnarly.
